### PR TITLE
[ReduceOp] Fix reduce op correctness issue when threads_per_warp < num_warps.

### PR DESCRIPTION
**Pull request by chengjunlu**
Fix the issue in `accumulatePartialReductions` when sizeInterWarps > threads_per_warp.
Use the warp reduction multiple times to get final result when the problem size > maximum reduce lane number.

### CODE REVIEW

File: third_party/intel/lib/TritonIntelGPUToLLVM/ReduceOpToLLVM.cpp [hunk 1]
```diff
@@ -310,48 +310,75 @@ struct ReduceOpConversion
...
-      }
+        threadsPerWarp;
+
+    // It is a batched reduce with the initial problem shape [elems /
+    // sizeInterWarps, sizeInterWarps]. The threadsPerWarp is 2^n. The
+    // sizeInterWarps is 2^m. With the horizontal warp reduction, the problem
+    // size is [elems / sizeInterWarps, N] -> [elems / sizeInterWarps, ceil(N,
+    // threadsPerWarp)] in each reduce iteration.
+    unsigned problemBatchSize = elems / sizeInterWarps;
+    for (unsigned problemSize = sizeInterWarps; problemSize > 0;
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/ReduceOpToLLVM.cpp:330]: how do we know sizeInterWarps is divisible by threadsPerWarp?

**Review by whitneywhtsang — COMMENTED** [review]
Can we upstream this change to OpenAI repo?

File: third_party/intel/lib/TritonIntelGPUToLLVM/ReduceOpToLLVM.cpp (hunk 1 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/ReduceOpToLLVM.cpp:330]: Good question.

I get this information at the comments at https://github.com/intel/intel-xpu-backend-for-triton/blob/7effcd094dae2b76dc05519df949cca37f02c93f/third_party/intel/lib/TritonIntelGPUToLLVM/ReduceOpToLLVM.cpp#L64

But I double checked the function https://github.com/intel/intel-xpu-backend-for-triton/blob/7effcd094dae2b76dc05519df949cca37f02c93f/lib/Analysis/Utility.cpp#L143

It seems no limitation in the sizeInterWarps should be 2^m.

It seems another common issue in ReduceOp conversion.

Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/ReduceOpToLLVM.cpp:330]: Let's add an assertion to ensure `sizeInterWarps%threadsPerWarp==0`.

**Comment by chengjunlu** [comment]
> Can we upstream this change to OpenAI repo?

We can try to upstream it.

File: third_party/intel/lib/TritonIntelGPUToLLVM/ReduceOpToLLVM.cpp (hunk 1 above)
Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/ReduceOpToLLVM.cpp:330]: Yes. 

If there is issue in the sizeInterWarps, I will fix it in a new PR.

**Review by whitneywhtsang — APPROVED** [review]
The change LGTM, but we should upstream it asap.

**Comment by etiotto** [comment]
> The change LGTM, but we should upstream it asap.

Do we have an issue to track upstreaming this? If not please create one for @chengjunlu.

**Pull request merged**